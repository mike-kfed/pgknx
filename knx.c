/*-------------------------------------------------------------------------
 *
 * knx.c
 *	  PostgreSQL type definitions for KNX Group/individual Addresses
 *
 * Author:	Michael Kefeder (htd)
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "fmgr.h"
#include "utils/builtins.h"

#include "knx.h"

PG_MODULE_MAGIC;

#ifdef USE_ASSERT_CHECKING
#define KNX_DEBUG 1
#else
#define KNX_DEBUG 0
#endif


enum knx_type
{
	INVALID, ANY, GA_LEVEL3, GA_LEVEL2, INDIVIDUAL
};

/*static const char *const knx_names[] = {"invalid", "KNX Address", "Group Address Level 3", "Group Address Level 2", "Individual Address"};*/

const size_t MAX_ADDR_LEN = 12;

/*----------------------------------------------------------
 * Formatting and conversion routines.
 *---------------------------------------------------------*/

static bool
string2group_addr(const char *addr, bool errorOK, knx_addr *result,
		   enum knx_type accept)
{
	unsigned int a=0, b=0, c=0, res;
	res = sscanf(addr, "%u/%u/%u", &a, &b, &c);
	if (res == 3 && a <= 0x1F && b <= 0x07 && c <= 0xFF
			&& (a > 0 || b > 0 || c > 0))
	{
		*result = (a << 11) | (b << 8) | c;
		return true;
	}
	if (res == 2 && a <= 0x1F && b <= 0x7FF
			&& (a > 0 || b > 0))
	{
		*result = (a << 11) | (b & 0x7FF);
		return true;
	}
	if (res == 1 && sscanf(addr, "%x", &a) == 1 && a <= 0xFFFF
			&& a > 0)
	{
		*result = a;
		return true;
	}
	ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("invalid group address format/values %s", addr)));
	return false;
}

static bool
string2individual_addr(const char *addr, bool errorOK, knx_addr *result)
{
	unsigned int a=0, b=0, c=0, res;
	res = sscanf(addr, "%u.%u.%u", &a, &b, &c);
	if (res == 3 && a == 0 && b == 0 && c == 0)
		goto kaputt;
	if (res == 3 && (a > 15 || b > 15 || c > 255))
		goto kaputt;
	if (res == 3)
	{
		*result = ((a & 0x0f) << 12) | ((b & 0x0f) << 8) | ((c & 0xff));
		return true;
	}
	res = sscanf(addr, "%x", &a);
	if (res == 1 && a > 0)
	{
		*result = a & 0xffff;
		return true;
	}
kaputt:
	ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("invalid individual address format/values %s", addr)));
	return false;
}

static bool
group_addr2string(knx_addr addr, bool errorOK, char *result)
{
	if (snprintf(result, MAX_ADDR_LEN, "%d/%d/%d",
			(addr >> 11) & 0x1f, (addr >> 8) & 0x07, (addr) & 0xff) < 0)
	{
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("cannot format to group address %d", addr)));
		return false;
	}
	return true;
}

static bool
group_addr2ga2string(knx_addr addr, bool errorOK, char *result)
{
	if (snprintf(result, MAX_ADDR_LEN, "%d/%d",
			(addr >> 11) & 0x1f, addr & 0x07ff) < 0)
	{
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("cannot format to group address %d", addr)));
		return false;
	}
	return true;
}

static bool
individual_addr2string(knx_addr addr, bool errorOK, char *result)
{
	if (snprintf(result, MAX_ADDR_LEN, "%d.%d.%d",
			(addr >> 12) & 0x0f, (addr >> 8) & 0x0f, (addr) & 0xff) < 0)
	{
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("cannot format to individual address %d", addr)));
		return false;
	}
	return true;
}

PG_FUNCTION_INFO_V1(knx_ga3_out);
Datum
knx_ga3_out(PG_FUNCTION_ARGS)
{
	knx_addr	val = PG_GETARG_KNX_ADDR(0);
	char	   *result;
	char		buf[MAX_ADDR_LEN + 1];

	(void) group_addr2string(val, false, buf);

	result = pstrdup(buf);
	PG_RETURN_CSTRING(result);
}

PG_FUNCTION_INFO_V1(knx_ga3_in);
Datum
knx_ga3_in(PG_FUNCTION_ARGS)
{
	const char *str = PG_GETARG_CSTRING(0);
	knx_addr		result;

	(void) string2group_addr(str, false, &result, GA_LEVEL3);
	PG_RETURN_KNX_ADDR(result);
}

PG_FUNCTION_INFO_V1(knx_ga2_out);
Datum
knx_ga2_out(PG_FUNCTION_ARGS)
{
	knx_addr	val = PG_GETARG_KNX_ADDR(0);
	char	   *result;
	char		buf[MAX_ADDR_LEN + 1];

	(void) group_addr2ga2string(val, false, buf);

	result = pstrdup(buf);
	PG_RETURN_CSTRING(result);
}

PG_FUNCTION_INFO_V1(knx_ga2_in);
Datum
knx_ga2_in(PG_FUNCTION_ARGS)
{
	const char *str = PG_GETARG_CSTRING(0);
	knx_addr		result;

	(void) string2group_addr(str, false, &result, GA_LEVEL2);
	PG_RETURN_KNX_ADDR(result);
}

PG_FUNCTION_INFO_V1(int_to_knx_addr);
Datum
int_to_knx_addr(PG_FUNCTION_ARGS)
{
	int32 arg1 = PG_GETARG_INT32(0);
	knx_addr		result;

	result = arg1;
	if ((uint16) result != arg1)
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("integer out of range")));
	if (result == 0)
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("address cannot be zero")));
	PG_RETURN_KNX_ADDR(result);
}

PG_FUNCTION_INFO_V1(knx_addr_to_int);
Datum
knx_addr_to_int(PG_FUNCTION_ARGS)
{
	knx_addr arg1 = PG_GETARG_KNX_ADDR(0);
	int32		result;

	result = arg1;
	if ((uint16) result != arg1)
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("integer out of range")));
	if (result == 0)
		ereport(ERROR,
			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
			 errmsg("address cannot be zero")));
	PG_RETURN_INT32(result);
}

/* Individual Addr */

PG_FUNCTION_INFO_V1(knx_ia_out);
Datum
knx_ia_out(PG_FUNCTION_ARGS)
{
	knx_addr	val = PG_GETARG_KNX_ADDR(0);
	char	   *result;
	char		buf[MAX_ADDR_LEN + 1];

	(void) individual_addr2string(val, false, buf);

	result = pstrdup(buf);
	PG_RETURN_CSTRING(result);
}

PG_FUNCTION_INFO_V1(knx_ia_in);
Datum
knx_ia_in(PG_FUNCTION_ARGS)
{
	const char *str = PG_GETARG_CSTRING(0);
	knx_addr		result;

	(void) string2individual_addr(str, false, &result);
	PG_RETURN_KNX_ADDR(result);
}
