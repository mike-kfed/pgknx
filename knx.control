# knx extension
comment = 'data types for KNX Addresses'
default_version = '1.0'
module_pathname = '$libdir/knx'
relocatable = true
