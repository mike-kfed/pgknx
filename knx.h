/*-------------------------------------------------------------------------
 *
 * knx.h
 *	  PostgreSQL type definitions for KNX Group/Internal Addresses
 *
 * Author:	Michael Kefeder (htd)
 *
 *-------------------------------------------------------------------------
 */

#ifndef KNX_H
#define KNX_H

#include "fmgr.h"

#undef KNX_DEBUG
#define KNX_WEAK_MODE

/*
 * uint16 is the internal storage format for KNX addresses.
 */
typedef uint16 knx_addr;

#define KNX_ADDR_FORMAT UINT16_FORMAT

#define PG_GETARG_KNX_ADDR(n) PG_GETARG_UINT16(n)
#define PG_RETURN_KNX_ADDR(x) PG_RETURN_UINT16(x)

extern void initialize(void);

#endif							/* KNX_H */
