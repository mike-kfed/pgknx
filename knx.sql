
-- complain if script is sourced in psql, rather than via CREATE EXTENSION
\echo Use "CREATE EXTENSION knx" to load this file. \quit

--
-- Input and output functions and data types:
--
---------------------------------------------------
-- Level 3
CREATE FUNCTION knx_ga3_in(cstring)
	RETURNS knx_group_address3
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ga3_out(knx_group_address3)
	RETURNS cstring
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ga3_recv(internal) RETURNS knx_group_address3
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_recv';
CREATE FUNCTION knx_ga3_send(knx_group_address3) RETURNS bytea
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_send';
CREATE TYPE knx_group_address3 (
	INPUT = knx_ga3_in,
	OUTPUT = knx_ga3_out,
	RECEIVE = knx_ga3_recv,
	SEND = knx_ga3_send,
	INTERNALLENGTH = 2,
	PASSEDBYVALUE,
	ALIGNMENT = int2
);
COMMENT ON TYPE knx_group_address3
	IS 'KNX Group Address Level 3';

CREATE FUNCTION int2ga3(integer) RETURNS knx_group_address3 IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'int_to_knx_addr';
CREATE CAST (integer AS knx_group_address3) WITH FUNCTION int2ga3(integer) AS IMPLICIT;

CREATE FUNCTION ga32int(knx_group_address3) RETURNS integer IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'knx_addr_to_int';
CREATE CAST (knx_group_address3 AS integer) WITH FUNCTION ga32int(knx_group_address3) AS IMPLICIT;

-- Level 2
CREATE FUNCTION knx_ga2_in(cstring)
	RETURNS knx_group_address2
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ga2_out(knx_group_address2)
	RETURNS cstring
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ga2_recv(internal) RETURNS knx_group_address2
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_recv';
CREATE FUNCTION knx_ga2_send(knx_group_address2) RETURNS bytea
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_send';
CREATE TYPE knx_group_address2 (
	INPUT = knx_ga2_in,
	OUTPUT = knx_ga2_out,
	RECEIVE = knx_ga2_recv,
	SEND = knx_ga2_send,
	INTERNALLENGTH = 2,
	PASSEDBYVALUE,
	ALIGNMENT = int2
);
COMMENT ON TYPE knx_group_address2
	IS 'KNX Group Address Level 2';

CREATE FUNCTION int2ga2(integer) RETURNS knx_group_address2 IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'int_to_knx_addr';
CREATE CAST (integer AS knx_group_address2) WITH FUNCTION int2ga2(integer) AS IMPLICIT;

CREATE FUNCTION ga32int(knx_group_address2) RETURNS integer IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'knx_addr_to_int';
CREATE CAST (knx_group_address2 AS integer) WITH FUNCTION ga32int(knx_group_address2) AS IMPLICIT;

-- Individual
CREATE FUNCTION knx_ia_in(cstring)
	RETURNS knx_individual_address
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ia_out(knx_individual_address)
	RETURNS cstring
	AS 'MODULE_PATHNAME'
	LANGUAGE C
	IMMUTABLE STRICT
	PARALLEL SAFE;
CREATE FUNCTION knx_ia_recv(internal) RETURNS knx_individual_address
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_recv';
CREATE FUNCTION knx_ia_send(knx_individual_address) RETURNS bytea
	IMMUTABLE
	STRICT
	LANGUAGE C
	AS 'MODULE_PATHNAME',
	'knx_addr_send';
CREATE TYPE knx_individual_address (
	INPUT = knx_ia_in,
	OUTPUT = knx_ia_out,
	RECEIVE = knx_ia_recv,
	SEND = knx_ia_send,
	INTERNALLENGTH = 2,
	PASSEDBYVALUE,
	ALIGNMENT = int2
);
COMMENT ON TYPE knx_individual_address
	IS 'KNX Individual Address';

CREATE FUNCTION int2ia(integer) RETURNS knx_individual_address IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'int_to_knx_addr';
CREATE CAST (integer AS knx_individual_address) WITH FUNCTION int2ia(integer) AS IMPLICIT;

CREATE FUNCTION ga32int(knx_individual_address) RETURNS integer IMMUTABLE STRICT LANGUAGE C AS '$libdir/knx', 'knx_addr_to_int';
CREATE CAST (knx_individual_address AS integer) WITH FUNCTION ga32int(knx_individual_address) AS IMPLICIT;
