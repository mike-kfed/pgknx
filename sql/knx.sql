--
-- Test KNX extension
--

CREATE EXTENSION knx;

-- Check whether any of our opclasses fail amvalidate
-- ... they will, because of missing cross-type operators
SELECT amname, opcname
FROM (SELECT amname, opcname, opc.oid
      FROM pg_opclass opc LEFT JOIN pg_am am ON am.oid = opcmethod
      WHERE opc.oid >= 16384
      ORDER BY 1, 2 OFFSET 0) ss
WHERE NOT amvalidate(oid);

--
-- test valid conversions
--
SELECT '17/6/1'::knx_group_address3;
SELECT '00FA'::knx_group_address3;
SELECT '0xA0FA'::knx_group_address3;
SELECT 165::knx_group_address3;

SELECT '29/2012'::knx_group_address2;
SELECT '02E7'::knx_group_address2;
SELECT 165::knx_group_address2;

SELECT '4.3.250'::knx_individual_address;
SELECT '12A5'::knx_individual_address;
SELECT 2000::knx_individual_address;

SELECT ('00FA'::knx_group_address3)::integer;
SELECT ('00FA'::knx_group_address2)::integer;
SELECT ('12A5'::knx_individual_address)::integer;
--
-- test invalid
--

SELECT '32/1/2'::knx_group_address3;
SELECT '0/0/0'::knx_group_address3;
SELECT '31/8/23'::knx_group_address3;
SELECT '31/1/300'::knx_group_address3;
SELECT '17/2050'::knx_group_address2;
SELECT '0000'::knx_group_address2;
SELECT '0x0000'::knx_group_address2;
SELECT '3300FA'::knx_group_address3;
SELECT 65000165::knx_group_address3;
SELECT 0::knx_group_address3;


SELECT 0::knx_individual_address;
SELECT '0.0.0'::knx_individual_address;
SELECT '16.0.1'::knx_individual_address;
SELECT '15.16.1'::knx_individual_address;
SELECT '15.12.256'::knx_individual_address;


--
-- cleanup
--
DROP EXTENSION knx;
