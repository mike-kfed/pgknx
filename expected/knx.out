--
-- Test KNX extension
--
CREATE EXTENSION knx;
-- Check whether any of our opclasses fail amvalidate
-- ... they will, because of missing cross-type operators
SELECT amname, opcname
FROM (SELECT amname, opcname, opc.oid
      FROM pg_opclass opc LEFT JOIN pg_am am ON am.oid = opcmethod
      WHERE opc.oid >= 16384
      ORDER BY 1, 2 OFFSET 0) ss
WHERE NOT amvalidate(oid);
 amname | opcname 
--------+---------
(0 rows)

--
-- test valid conversions
--
SELECT '17/6/1'::knx_group_address3;
 knx_group_address3 
--------------------
 17/6/1
(1 row)

SELECT '1/0/0'::knx_group_address3;
 knx_group_address3 
--------------------
 1/0/0
(1 row)

SELECT '0/1/0'::knx_group_address3;
 knx_group_address3 
--------------------
 0/1/0
(1 row)

SELECT '0/0/1'::knx_group_address3;
 knx_group_address3 
--------------------
 0/0/1
(1 row)

SELECT '00FA'::knx_group_address3;
 knx_group_address3 
--------------------
 0/0/250
(1 row)

SELECT '0xA0FA'::knx_group_address3;
 knx_group_address3 
--------------------
 20/0/250
(1 row)

SELECT 165::knx_group_address3;
 knx_group_address3 
--------------------
 0/0/165
(1 row)

SELECT '29/2012'::knx_group_address2;
 knx_group_address2 
--------------------
 29/2012
(1 row)

SELECT '1/0'::knx_group_address2;
 knx_group_address2 
--------------------
 1/0
(1 row)

SELECT '0/1'::knx_group_address2;
 knx_group_address2 
--------------------
 0/1
(1 row)

SELECT '02E7'::knx_group_address2;
 knx_group_address2 
--------------------
 0/743
(1 row)

SELECT 165::knx_group_address2;
 knx_group_address2 
--------------------
 0/165
(1 row)

SELECT '4.3.250'::knx_individual_address;
 knx_individual_address 
------------------------
 4.3.250
(1 row)

SELECT '12A5'::knx_individual_address;
 knx_individual_address 
------------------------
 1.2.165
(1 row)

SELECT '0x1145'::knx_individual_address;
 knx_individual_address 
------------------------
 1.1.69
(1 row)

SELECT 2000::knx_individual_address;
 knx_individual_address 
------------------------
 0.7.208
(1 row)

--
-- test casts
--
SELECT ('00FA'::knx_group_address3)::integer;
 int4 
------
  250
(1 row)

SELECT ('00FA'::knx_group_address2)::integer;
 int4 
------
  250
(1 row)

SELECT ('12A5'::knx_individual_address)::integer;
 int4 
------
 4773
(1 row)

--
-- test invalid
--
SELECT '32/1/2'::knx_group_address3;
ERROR:  invalid group address format/values 32/1/2
LINE 1: SELECT '32/1/2'::knx_group_address3;
               ^
SELECT '0/0/0'::knx_group_address3;
ERROR:  invalid group address format/values 0/0/0
LINE 1: SELECT '0/0/0'::knx_group_address3;
               ^
SELECT '31/8/23'::knx_group_address3;
ERROR:  invalid group address format/values 31/8/23
LINE 1: SELECT '31/8/23'::knx_group_address3;
               ^
SELECT '31/1/300'::knx_group_address3;
ERROR:  invalid group address format/values 31/1/300
LINE 1: SELECT '31/1/300'::knx_group_address3;
               ^
SELECT '3300FA'::knx_group_address3;
ERROR:  invalid group address format/values 3300FA
LINE 1: SELECT '3300FA'::knx_group_address3;
               ^
SELECT 65000165::knx_group_address3;
ERROR:  integer out of range
SELECT 0::knx_group_address3;
ERROR:  address cannot be zero
SELECT '17/2050'::knx_group_address2;
ERROR:  invalid group address format/values 17/2050
LINE 1: SELECT '17/2050'::knx_group_address2;
               ^
SELECT '0000'::knx_group_address2;
ERROR:  invalid group address format/values 0000
LINE 1: SELECT '0000'::knx_group_address2;
               ^
SELECT '0x0000'::knx_group_address2;
ERROR:  invalid group address format/values 0x0000
LINE 1: SELECT '0x0000'::knx_group_address2;
               ^
SELECT 0::knx_individual_address;
ERROR:  address cannot be zero
SELECT '0.0.0'::knx_individual_address;
ERROR:  invalid individual address format/values 0.0.0
LINE 1: SELECT '0.0.0'::knx_individual_address;
               ^
SELECT '0x0'::knx_individual_address;
ERROR:  invalid individual address format/values 0x0
LINE 1: SELECT '0x0'::knx_individual_address;
               ^
SELECT '16.0.1'::knx_individual_address;
ERROR:  invalid individual address format/values 16.0.1
LINE 1: SELECT '16.0.1'::knx_individual_address;
               ^
SELECT '15.16.1'::knx_individual_address;
ERROR:  invalid individual address format/values 15.16.1
LINE 1: SELECT '15.16.1'::knx_individual_address;
               ^
SELECT '15.12.256'::knx_individual_address;
ERROR:  invalid individual address format/values 15.12.256
LINE 1: SELECT '15.12.256'::knx_individual_address;
               ^
--
-- test op classes
--
create table knx_test(
	a knx_group_address3 primary key,
	b knx_group_address2,
	c knx_individual_address
);
create index on knx_test(b);
create index on knx_test(c);
drop table knx_test;
--
-- cleanup
--
DROP EXTENSION knx;
